mypy/semanal.py

```diff
@@ -581,7 +581,11 @@ def refresh_partial(
     def refresh_top_level(self, file_node: MypyFile) -> None:
         """Reanalyze a stale module top-level in fine-grained incremental mode."""
         self.recurse_into_functions = False
-        self.add_implicit_module_attrs(file_node)
+        # We do it in the very last order, because of
+        # `builtins.dict <-> typing.Mapping <-> abc.ABCMeta`
+        # cyclic imports.
+        if file_node.fullname not in ("typing", "abc", "builtins"):
```

Comment:
I would instead re-use `core_modules` (defined in `semanal_main.py`) here and in other place. Also you may need to use a flag for this method to avoid an import cycle with the latter (like `skip_implicit_attrs`).